### PR TITLE
GSSPRT-123: Unlink Contribution Action

### PR DESCRIPTION
**Pull request by deb1990**
## Overview
As part of this PR, a new Case action "Unlink Contribution" has been added. It is shown only when the Prospect is converted to `contribution`. This action, moves the contribution to "cancelled" status, and also unlinks the contribution from the prospect.

## Before
No such action was present.

## After
2021-05-27 at 11 22 AM

## Technical Details
In `ang/prospect/app.config.js`, added the new action.
Added `ang/prospect/case-actions/services/unlink-contribution-case-action.service.js` service, to control the visibility of the action.

### CODE REVIEW

File: ang/prospect/case-actions/services/unlink-contribution-case-action.service.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+     */
+    this.isActionAllowed = function (action, cases) {
+      if (!cases[0] || !cases[0].prospect.paymentInfo) {
+        return;
+      }
+
+      var isContributeTypeProspect =
+        cases[0].prospect.paymentInfo.payment_entity === 'contribute';
+
+      return cases[0] &&
```
Comment (lisandro-compucorp) [inline ang/prospect/case-actions/services/unlink-contribution-case-action.service.js]: I think at this stage the `cases[0]` has no chance to be false, then it would be unnecessary to evaluate it.

File: ang/prospect/case-actions/services/unlink-contribution-case-action.service.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+     * @param {object} action action
+     * @param {Array} cases cases
+     * @returns {boolean} if action is allowed
+     */
+    this.isActionAllowed = function (action, cases) {
+      if (!cases[0] || !cases[0].prospect.paymentInfo) {
+        return;
+      }
+
+      var isContributeTypeProspect =
```
Comment (lisandro-compucorp) [inline ang/prospect/case-actions/services/unlink-contribution-case-action.service.js:22]: could we do an early return here? I mean:
```js
if (!isContributeTypeProspect) {
  return false;
}
```
and forget of the variable later. I try to avoid the long conditions like here, but it is just a suggestion :)

File: ang/prospect/case-actions/services/unlink-contribution-case-action.service.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    /**
+     * Checks if the Action is allowed
+     *
+     * @param {object} action action
+     * @param {Array} cases cases
+     * @returns {boolean} if action is allowed
+     */
+    this.isActionAllowed = function (action, cases) {
+      if (!cases[0] || !cases[0].prospect.paymentInfo) {
+        return;
```
Comment (lisandro-compucorp) [inline ang/prospect/case-actions/services/unlink-contribution-case-action.service.js:19]: This maybe is my lack of javascript knowledge, but let me ask: this is exactly the same as putting `return false`? I would prefer the latter.

File: ang/prospect/prospect-converted/services/prospect-converted.service.js [hunk 4]
```diff
@@ -12,17 +12,17 @@
    */
   function ProspectConverted (crmApi, ProspectGlobalValues, CaseTypeCategory) {
     /**
-     * Returns if the case is converted to Prospect
+     * Returns if the prospect converted value
```
Comment (lisandro-compucorp) [inline ang/prospect/prospect-converted/services/prospect-converted.service.js]: Just a typo: the `if` is not required anymore.

File: ang/prospect/case-actions/services/unlink-contribution-case-action.service.js (hunk 1 above)
Comment (deb1990) [reply ang/prospect/case-actions/services/unlink-contribution-case-action.service.js]: No it can be false, for case bulk actions. Before you select any cases, this array would be empty. So need to check.

Comment (lisandro-compucorp) [reply ang/prospect/case-actions/services/unlink-contribution-case-action.service.js]: yes, but I mean, it was checked a few lines before, on this method

Comment (deb1990) [reply ang/prospect/case-actions/services/unlink-contribution-case-action.service.js]: Oh Okay, yep I removed that already :D (while fixing other comments), thanks. I thought you are talking about the first line.

File: ang/prospect/case-actions/services/unlink-contribution-case-action.service.js (hunk 3 above)
Comment (deb1990) [reply ang/prospect/case-actions/services/unlink-contribution-case-action.service.js:19]: Yes, `return` or `return false` or `return undefined`, all are same in Javascript. And because Javascript is not a typed language, it is a common practise to just write `return`. Also the same approach is heavily used in all CPS products.
So while I agree with you, that `return false` is better. But on the other hand, its an opinionated view, which to be honest most JS developers do not follow.
And we have a style guide called `semistandard` that we use for JS linting, and it does not enforce https://eslint.org/docs/rules/consistent-return rule. Hence I would say lets leave this to the choice of the developer who is writing the code. If someone chooses to mention consistent return type, all good, and if they dont, thats also good.

TLDR: This is an opinionated view, and currently the JS linter used in this project does not enforce this.

Comment (lisandro-compucorp) [reply ang/prospect/case-actions/services/unlink-contribution-case-action.service.js:19]: thanks for the explanation! I understand now

**Review by lisandro-compucorp — DISMISSED** [review]
👍

**Review by tunbola — APPROVED** [review]
